### ISSUE TITLE
Feature: Context aware rendering.

### ISSUE THREAD

**Issue by marblewraith** [scottwillmoore/obsidian-short-links#7]
### Example:

Lets say there's a note `test` with heading `## Germane cognitive load` in it.

Then another note `linked` that references `test#Germane cognitive load`.

When using this plugin, the `linked` note reads as follows:

"Foo bar baz, this is blah relating to `Germane cognitive load` and will proceed at 08:00."

### Problem:

In the original `test` note, Germane is capitalized because it's a heading. But in the `linked` note there is no need for Germane to be capitalized. That is, it's in the middle of a sentence, it's not a noun, etc. This would be even worse for someone using title case.

### Solution:

- Detect whether link starts after a newline char, or period-space (`. `), if so display as "Sentence case", else lowercase.

- Create an inline override so that this context awareness can be selectively overridden. For example maybe if you did `|+` on the end of the link, it should display verbatim as it is originally written.

**Comment by scottwillmoore** [comment]
Yes, I agree. I believe I have had this exact same problem a while back, but I can't remember my opinion on the situation.

I was able to find a few related discussions and feature requests on Reddit and the Obsidian Forum.

- [Top tip: links aren't case-sensitive](https://www.reddit.com/r/ObsidianMD/comments/sirpri/top_tip_links_arent_casesensitive/)
- [Backlinking to page titles case-insensitive](https://forum.obsidian.md/t/backlinking-to-page-titles-case-insensitive/2844)
- [Maintain casing of text when using link autosuggestion](https://forum.obsidian.md/t/maintain-casing-of-text-when-using-link-autosuggestion/23041)

## Solution

### Use aliases

`[[Cognitive load#Germane cognitive load|germane cognitive load]]`

👍 It does work right now. No plugin is required.
👍 You can handle any situation.

👍👎 The alias will **not** be updated when you rename notes and headings.

👎 It is tedious to create an alias for each link.

### Use case-insensitive links

As discussed on Reddit and the Obsidian Forum, internal links are case-insensitive.

`[[Cognitive load#germane cognitive load]]`

👍 It does work right now. No plugin is required.

👍👎 The link will be updated when you rename notes and headings.

👎 You can only handle situations related to case.
👎 Auto-complete will try to match the original case.
👎 When you rename notes and headings, all links will be updated to match their original case.

### Use context-aware renderer

`[[Cognitive load#Germane cognitive load]]`

👍 It will be automatic, and require no additional effort.

👎 It does **not** work right now. A plugin is required.
👎 It will be **very** difficult to handle all edge cases.

To add to the last point, I think it is important to discuss all the edge cases. I think this is the most important, but also the most difficult component of this problem.

🔗 `[[Cognitive load#Germane cognitive load]]`
"*Germane cognitive load* is the processing, construction and automation of schemas." 
"The third type of load that has been proposed is *germane cognitive load*."

🔗 `[[Australia]]`
"The state of Victoria is on the east coast of *Australia*." (Not *australia*).
"He was an *Australian* man." (Not *australia*).
"The were more *Australians* than any other nationality." (Not *australia*).

🔗 `[[University of Colorado]]`
"... the north edge of the *University of Colorado* campus". (Not *university of Colorado* or *university of colorado*).

Above are some examples I could think of in a short amount of time, but I am sure there are many, many more. So I think your current solution (a context aware renderer) might just create more problems than solutions. I have not even considered [apostrophe rules](https://www.grammarly.com/blog/apostrophe/) or non-English languages. This problem at the extreme starts to become [natural language processing](https://en.wikipedia.org/wiki/Natural_language_processing) which makes it very difficult to create a good solution.

## Conclusion

We could agree to just solve a subset of these problems, such as just fix problems related to case and ignore problems related to pluralization. However, this could end up just being messy and create more confusion. Therefore my current stance is that aliases, although ugly and tedious might still be the most robust solution.

Maybe a different approach is to investigate improvements to the auto-complete and provide tools to allow users that rename notes and headings to easily double check and manually amend all links with aliases. I would love to hear any other solutions or thoughts that people might have on this problem.
